enhancements/olm/max-openshift-versions-for-operators.md

```diff
@@ -0,0 +1,284 @@
+
+Many operators shipped by Red Hat are rigorously tested on a specific set of OpenShift versions. The teams responsible for this testing know exactly which `{major}.{minor}` versions of OpenShift that their operators can run on. OLM should allow Operator Authors to specify the maximum `{major}.{minor}` version of OpenShift that their operator may run on.
+
+With this information OLM should:
+
+- Prevent Minor OpenShift Upgrades when it is possible to determine that one or more installed operators will not run on the next minor OpenShift version.
+- Prevent the operator from being installed on OpenShift clusters whose version is greater than the `maxOpenShiftVersion` specified by the operator.
+
+## Motivation
+
+The primary purpose of this enhancement is to define how OpenShift can protect existing services introduced by OLM Managed Operators when performing Minor OpenShift Upgrades.
+
+### Goals
+
+- Prevent Minor OpenShift Upgrades when an installed operator declares that it does not support the next minor OpenShift version.
+- Warn Cluster Admins when an installed operator does not explicitly declare support for the next minor OpenShift version.
+- Prevent OLM from installing operators whose `maxOpenShiftVersion` is less than the current [ClusterVersion](https://github.com/openshift/api/blob/a9e731090f5ed361e5ab887d0ccd55c1db7fc633/config/v1/types_cluster_version.go#L11-L13).
+
+### Non-Goals
+
+- Ensure that services will not be disrupted from a cluster upgrade.
+- Guarantee that all installed OLM Managed Operators will continue to run after a Minor OpenShift Upgrade.
+- Prevent Patch OpenShift Upgrades.
+
+## Proposal
+
+The [ClusterVersionOperator](https://github.com/openshift/cluster-version-operator) provides OLM with the means to prevent Minor OpenShift Upgrades by setting the `operator-lifecycle-manager's` [ClusterOperator's upgradeable condition](https://github.com/openshift/api/blob/5935a5beec4bb8e1e81dd0fe9ebc2af36b9a09ae/config/v1/types_cluster_operator.go#L171-L175) to `False`.
+
+OLM currently reports that it is upgradeable as soon its deployments are successfully rolled out. OLM should instead set its `upgradeable condition` to reflect if any OLM Managed Operators indicate that they will not run on the next minor OpenShift version.
+
+The bulk of this enhancement focuses on defining:
+
+- How operators will define their `maxOpenShiftVersion`.
+- The steps that OLM will take to determine upgrade safety based on the collection of installed operators.
+- The steps that OLM will take to prevent the operator from being installed on OpenShift clusters whose version is greater than the `maxOpenShiftVersion` specified by the operator.
+
+### Allowing Operator Authors to Define a Maximum OpenShift Version
```

Comment:
Can I specify both `minKubeversion`and `maxOpenShiftVersion`? If yes, what happens if `minKubeversion` is `v0.19` and `maxOpenShiftVersion` is `v4.6`? I suggest you call out the various permutations / combinations and the expected result.

Reply:
they're independent constraints, I would expect both of them must be met to install, along with any other constraints.